internal/metrics/directory_usage_collector.go

```diff
@@ -0,0 +1,69 @@
+package metrics
+
+import (
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/sirupsen/logrus"
+)
+
+const (
+	directoryUsageUsedBytesName        = "assisted_installer_directory_usage_bytes"
+	directoryUsageUsedBytesDescription = "The amount of the free space in a directory that has been used in bytes"
+	fsUsageFreeBytesName               = "assisted_installer_fs_free_bytes"
+	fsUsageFreeBytesDescription        = "The amount of free space that is remaining for the filesystem backing this directory"
+
+	directoryLabelKey = "directory"
+)
+
+type directoryUsageCollectorConfig struct {
+	directories []string
+}
+
+func NewDirectoryUsageCollectorConfig(directories []string) *directoryUsageCollectorConfig {
+	return &directoryUsageCollectorConfig{directories: directories}
+}
+
+type directoryUsageCollector struct {
+	// The directory for which to report disk usage stats
+	directories              []string
+	diskStatsHelper          DiskStatsHelper
+	directoryUsageBytesDescs map[string]*prometheus.Desc
+	fsFreeByteDescs          map[string]*prometheus.Desc
+	log                      *logrus.Logger
+}
+
+func newDirectoryUsageCollector(directories []string, diskStatsHelper DiskStatsHelper, log *logrus.Logger) *directoryUsageCollector {
+	directoryUsageBytesDescs := make(map[string]*prometheus.Desc)
+	fsFreeByteDescs := make(map[string]*prometheus.Desc)
+	for _, directory := range directories {
+		directoryUsageBytesDescs[directory] = prometheus.NewDesc(directoryUsageUsedBytesName, directoryUsageUsedBytesDescription, nil, map[string]string{directoryLabelKey: directory})
+		fsFreeByteDescs[directory] = prometheus.NewDesc(fsUsageFreeBytesName, fsUsageFreeBytesDescription, nil, map[string]string{directoryLabelKey: directory})
+	}
+	return &directoryUsageCollector{
+		directories:              directories,
+		diskStatsHelper:          diskStatsHelper,
+		directoryUsageBytesDescs: directoryUsageBytesDescs,
+		fsFreeByteDescs:          fsFreeByteDescs,
+		log:                      log,
+	}
+}
+
+func (c *directoryUsageCollector) Describe(ch chan<- *prometheus.Desc) {
+	for _, desc := range c.directoryUsageBytesDescs {
```

Comment:
As `directoryUsageBytesDescs` and `fsFreeByteDescs` are maps we shouldn't rely on its order to do this, as golang randomizes access.

Reply:
I'm not depending on order, just expecting a sequence of directories, the order of that sequence is unimportant. Only that they are all handled.

Reply:
Also, we are building a map here, the key of the map is the directory, which we address later in the same way.

At no point do we depend on the order of the map.